flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java

```diff
@@ -0,0 +1,451 @@
+            Executing exec = getExecutingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == exec);
+
+            ctx.setExpectFinished(
+                    archivedExecutionGraph -> {
+                        assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED));
+                    });
+            exec.onEnter();
+            exec.suspend(new RuntimeException("suspend"));
+        }
+    }
+
+    @Test
+    public void testScaleUp() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+
+            ctx.setExpectRestarting(
+                    restartingArguments -> {
+                        // expect immediate restart on scale up
+                        assertThat(restartingArguments.getBackoffTime(), is(Duration.ZERO));
+                    });
+            ctx.setCanScaleUp(() -> true);
+            exec.notifyNewResourcesAvailable();
+        }
+    }
+
+    @Test
+    public void testNoScaleUp() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            ctx.setCanScaleUp(() -> false);
+            exec.onEnter();
+            exec.notifyNewResourcesAvailable();
+        }
+    }
+
+    @Test
+    public void testFailingOnDeploymentFailure() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setCanScaleUp(() -> false);
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+            ctx.setExpectFailing(assertNonNull());
+
+            // create ExecutionGraph with one ExecutionVertex, which fails during deployment.
+            JobGraph jobGraph = new JobGraph(new JobVertex("test"));
+            Executing exec = getExecutingState(ctx, null, jobGraph);
+            TestingLogicalSlotBuilder slotBuilder = new TestingLogicalSlotBuilder();
+            TaskManagerGateway taskManagerGateway =
+                    new ExecutionVertexDeploymentTest.SubmitFailingSimpleAckingTaskManagerGateway();
+            slotBuilder.setTaskManagerGateway(taskManagerGateway);
+            LogicalSlot slot = slotBuilder.createTestingLogicalSlot();
+            exec.getExecutionGraph()
+                    .getAllExecutionVertices()
+                    .forEach(executionVertex -> executionVertex.tryAssignResource(slot));
+
+            // trigger deployment
+            exec.onEnter();
+        }
+    }
+
+    @Test
+    public void testFailureReportedViaUpdateTaskExecutionState() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ExecutionGraph returnsFailedStateExecutionGraph =
+                    new ReturnsTrueOnUpdateExecutionGraph();
+            Executing exec = getExecutingState(ctx, returnsFailedStateExecutionGraph, null);
```

Comment:
Instead of passing null arguments, it is usually a better idea to give a more expressive method name w/o this parameter.